reinitialize-pods/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+use futures_util::StreamExt;
+use k8s_openapi::api::core::v1::Pod;
+use kube::runtime::watcher;
+use kubert::Runtime;
+
+// ERRNO 95: Operation not supported
+pub const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+const DATA_PLANE_LABEL: &str = "linkerd.io/control-plane-ns";
+const CONDITION_EVICTED_REASON: &str = "EvictionByEvictionAPI";
+
+pub fn run(runtime: &mut Runtime, node_name: String) {
+    let client = runtime.client().clone();
+    let pod_evts = runtime.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+
+    tokio::spawn(async move {
+        tokio::pin!(pod_evts);
+        while let Some(evt) = pod_evts.next().await {
+            if let watcher::Event::Applied(pod) = evt {
+                let maybe_terminated = pod
+                    .status
+                    .clone()
+                    .and_then(|x| x.init_container_statuses)
+                    .and_then(|x| x.into_iter().next())
+                    .filter(|x| x.name == "linkerd-network-validator")
+                    .and_then(|x| x.last_state)
+                    .and_then(|x| x.terminated)
+                    .filter(|x| x.exit_code == UNSUCCESSFUL_EXIT_CODE);
+
+                let maybe_already_evicting =
+                    pod.status.clone().and_then(|x| x.conditions).and_then(|x| {
+                        x.into_iter().find(|y| {
+                            y.reason
+                                .as_ref()
+                                .is_some_and(|z| z == CONDITION_EVICTED_REASON)
+                        })
+                    });
+
```

Comment:
I think we could potentially simplify this if we return early at the start when the status is missing? We don't expect that to happen often anyway, but as a precaution we could continue iterating over events:

```rust
if let watcher::Event::Applied(pod) = evt {
                let status = if let Some(ref status) = pod.status {
                    status.clone()
                } else {
                    tracing::info!("Skipped, no status");
                    continue;
                };
```

I have also tried to restructure the `maybe_terminated` and `maybe_already_evicting` lookups to avoid clones and to return a bool:

```rust
             let terminated = status
                    .init_container_statuses
                    .unwrap_or_default()
                    .iter()
                    .find(|x| x.name == "linkerd-network-validator")
                    .and_then(|x| x.last_state.as_ref())
                    .and_then(|x| x.terminated.as_ref())
                    .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
                    .unwrap_or(false);

                let evicted = status.conditions.as_ref().and_then(|xs| {
                    xs.iter().find(|y| {
                        y.reason
                            .as_ref()
                            .is_some_and(|z| z == CONDITION_EVICTED_REASON)
                    })
                });
```

We could probably use `find()` in terminated to operate on an option instead of using the filter adapter. We know we will have at most one validator. The rest is similar to your solution. I'm not adamant on my solution, I think yours reads well too  but leaving here in case it's helpful :)

Reply:
Nice suggestions, that reads better indeed :+1: